### ISSUE TITLE
PCIe V100 (no NVLink): three bootstrap/runtime blockers + fixes (NCCL P2P hang, custom all-reduce livelock at graph capture, missing ninja)

### ISSUE THREAD

**Issue by Redhatvale** [dnv2003/v100-skinny#3]
# Draft issue for dnv2003/v100-skinny — review before posting

**Title:** PCIe V100 (no NVLink): three bootstrap/runtime blockers + fixes (NCCL P2P hang, custom all-reduce livelock at graph capture, missing ninja)

## Environment

- 2× Tesla V100-PCIE-32GB (SM70), PCIe, **no NVLink** (topology: PXB/PIX through PCIe bridges)
- Ubuntu 24.04, driver 580.173.02, CUDA toolkit 12.8
- v100-skinny @ master + 1Cat-vLLM 1.2.2 (pinned wheel, SHA256 verified)
- Checkpoint RadixArk/Qwen3.8-27B-NVFP4 @ 554ebba9, TP2

All published numbers in the repo are SXM2+NVLink; on PCIe the stock
instructions hit three reproducible blockers. Fixes below; with all three,
the stack boots in ~2.5 min and serves stably.

## Blocker 1 — JIT kernel build fails: `Ninja is required`

`torch.utils.cpp_extension.load()` shells out to the `ninja` **binary**;
installing the Python module alone is not enough — the venv's `bin/` must be
on `PATH`. Fix:

```bash
.venv-sm70/bin/pip install ninja
export PATH="$PWD/.venv-sm70/bin:$PATH"
```

Suggest: add `ninja` to requirements and a PATH note in bootstrap-sm70.sh
(the warm-up step fails silently into the server's boot wait otherwise).

## Blocker 2 — NCCL init hangs on PCIe (no NVLink)

Server stalls right after `vLLM is using nccl==2.27.5`; workers never finish
`ncclCommInitRank`. Fix:

```bash
NCCL_P2P_DISABLE=1 NCCL_IB_DISABLE=1
```

(P2P over these PCIe bridges doesn't complete; SHM transport works fine.
Decode perf is unaffected; prefill drops as expected without NVLink.)

## Blocker 3 — CUDA graph capture livelocks (custom all-reduce on PCIe)

With NCCL fixed, boot proceeds to `Capturing CUDA graphs` and then **spins
forever**: both GPUs pinned at 100 % util, workers at ~100 % CPU, log stops
after the Skinny route map / `QPN8_CENSUS_RUN` lines; only the 60 s
`shm_broadcast` warnings repeat. Left for 60+ minutes, never finishes.

Fix: start with **`--disable-custom-all-reduce`**. Boot then completes in
~2.5 min (with warm caches). Presumably vLLM's custom P2P all-reduce
spin-waits on peer flags that never land over this PCIe topology.

Suggest: auto-disable custom all-reduce when P2P is unavailable, or document
it next to the TP2 notes.

## Results with the three fixes (PCIe, TP2, k=7, KV fp8_e5m2, GMU 0.9)

- llama-benchy: tg256 = **61 avg / 74 peak tok/s**, pp1024 = 476 tok/s
- MTP mean acceptance: 5.0 (greedy) / 2.4–2.8 (sampling, essay prompts)
- `scalar_paged` = 0, QPN2/QPN8 routes active, smoke test coherent
- Reference on the same box: GGUF Q8_0 + llama.cpp + DFlash2 ≈ 43 tok/s

Happy to test patches. Thanks for the excellent kernels — Volta lives on.

**Comment by dnv2003** [comment]
Will fix ninja bug, I'd drop `--kv-cache-dtype fp8_e5m2`: it costs +4.82 ms/round on this model as it changes how attention is routed (will also look into fixing this) and finally graph capture failing is likely a symptom of p2p not being enabled, enabling p2p would allow for reduced round latency and improve decode.

I'll rent a pcie box to try to figure out the majority of the kinks since this repo has only been tested on the sxm2 server I have at home.
